implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java

```diff
@@ -158,57 +310,65 @@ private final SmallRyeHealth getHealth(Iterable<HealthCheck>... checks) {
         return new SmallRyeHealth(builder.build());
     }
 
-    private HealthCheckResponse.State processChecks(Iterable<HealthCheck> checks, JsonArrayBuilder results,
-            HealthCheckResponse.State status) {
-        if (checks != null) {
-            for (HealthCheck check : checks) {
-                status = fillCheck(check, results, status);
-            }
-        }
-
-        return status;
-    }
-
-    private HealthCheckResponse.State fillCheck(HealthCheck check, JsonArrayBuilder results,
+    private HealthCheckResponse.State handleResponse(HealthCheckResponse response, JsonArrayBuilder results,
             HealthCheckResponse.State globalOutcome) {
-        if (check == null) {
-            return globalOutcome;
-        }
-        JsonObject each = jsonObject(check);
-        results.add(each);
+        JsonObject responseJson = jsonObject(response);
+        results.add(responseJson);
+
         if (globalOutcome == HealthCheckResponse.State.UP) {
-            String status = each.getString("status");
+            String status = responseJson.getString("status");
             if (status.equals("DOWN")) {
                 return HealthCheckResponse.State.DOWN;
             }
         }
+
         return globalOutcome;
     }
 
-    private JsonObject jsonObject(HealthCheck check) {
-        try {
-            return jsonObject(check.call());
-        } catch (RuntimeException e) {
-            // Log Stacktrace to server log so an error is not just in Health Check response
-            HealthLogging.log.healthCheckError(e);
-
-            HealthCheckResponseBuilder response = HealthCheckResponse.named(check.getClass().getName()).down();
-
-            if (null != uncheckedExceptionDataStyle) {
-                switch (uncheckedExceptionDataStyle) {
-                    case ROOT_CAUSE:
-                        response.withData(ROOT_CAUSE, getRootCause(e).getMessage());
-                        break;
-                    case STACK_TRACE:
-                        response.withData(STACK_TRACE, getStackTrace(e));
-                        break;
-                    default:
-                        // don't add anything
-                }
-            }
+    private Uni<HealthCheckResponse> callAsync(AsyncHealthCheck asyncHealthCheck) {
+        // must be a supplier to be reinvoked on every subcribtion, don't change to method reference
+        return withRecovery(asyncHealthCheck.getClass().getName(),
+                Uni.createFrom().deferred(() -> {
+                    try {
+                        Uni<HealthCheckResponse> call = asyncHealthCheck.call();
+                        if (call == null) {
+                            throw HealthMessages.msg.healthCheckNull();
+                        }
+                        return call;
+                    } catch (IllegalArgumentException ex) {
+                        if (ex.getMessage().matches(".*supplier.*null.*")) {
```

Comment:
Which exception are you trying to catch here?

Reply:
java.lang.IllegalArgumentException: `supplier` must not be `null`

Reply:
I added a tests with the following checks - https://github.com/smallrye/smallrye-health/pull/111/files#diff-0e8e672ac11f35e6ee999ca9f6eeb5d2R79-R98. I don't like this exception handling but I can't find a way how to propagate supplier null exception through the uni itself.

Reply:
I would just rethrow the exception. 

That exception means that the Uni creation is invalid (`Uni.createFrom().item((Supplier) null);`).

Reply:
ok, that will be easier from the handling perspective. I wanted to have null handling unified but we can go back to this handling if requested by users.